Handle more cloud errors

andrew4699 opened the pull request:
<!--
    Possible security vulnerabilities: STOP here and contact security@apache.org instead!

    Please update the title of the PR with a meaningful message - do not leave it "empty" or "generated"
    Please update this summary field:

    The summary should cover these topics, if applicable:
    * the motivation for the change
    * a description of the status quo, for example the current behavior
    * the desired behavior
    * etc

    PR checklist:
    - Do a self-review of your code before opening a pull request
    - Make sure that there's good test coverage for the changes included in this PR
    - Run tests locally before pushing a PR (./gradlew check)
    - Code should have comments where applicable. Particularly hard-to-understand
      areas deserve good in-line documentation.
    - Include changes and enhancements to the documentation (in site/content/in-dev/unreleased)
    - For Work In Progress Pull Requests, please use the Draft PR feature.

    Make sure to add the information BELOW this comment.
    Everything in this comment will NOT be added to the PR description.
-->

Currently most cloud (AWS, Azure, GCP) errors are returned to users as 500s, even when the error is most likely "user-inflicted". This makes it difficult to monitor for real Polaris-side issues.

In a previous PR I fixed this for unauthorized errors. In this PR, I go one step further to extract the status code from the exception itself and forward it as the Polaris status code.

andrew4699 commented on quarkus/service/src/test/java/org/apache/polaris/service/quarkus/IcebergExceptionMapperTest.java:
```diff
@@ -16,11 +16,15 @@
  * under the License.
  */
-package org.apache.polaris.service.dropwizard;
+package org.apache.polaris.service.quarkus;
```
It was already in this package, just fixed the declaration.

dimas-b commented on quarkus/service/src/test/java/org/apache/polaris/service/quarkus/IcebergExceptionMapperTest.java:
```diff
@@ -52,4 +68,17 @@ void fileIOExceptionMapping(RuntimeException ex, int statusCode) {
+   * @param statusCode
+   * @return
+   */
+  private static HttpResponse mockAzureResponse(int statusCode) {
+    HttpResponse res = mock(HttpResponse.class);
+    when(res.getStatusCode()).thenReturn(statusCode);
+    return res;
+  }
```
What makes it specific to Azure?

andrew4699 replied:
Azure has its own `com.azure.core.http.HttpResponse` class.

dimas-b replied:
I see. Sorry, I missed that :sweat_smile:

dimas-b commented on service/common/src/main/java/org/apache/polaris/service/exception/IcebergExceptionMapper.java:
```diff
@@ -144,4 +112,56 @@ public static boolean containsAnyAccessDeniedHint(String message) {
+
+  static int mapExceptionToResponseCode(RuntimeException rex) {
+    // Cloud exceptions
+    if (rex instanceof S3Exception
+        || rex instanceof AzureException
+        || rex instanceof StorageException) {
```
Why bother with complex logic when we have a typed switch below?

andrew4699 replied:
Do you mean the 1st or 2nd switch? This guards the `doesAnyThrowableContainAccessDeniedHint` check, preserving the original behavior. I'm hoping to limit any unintended consequences by scoping this down to cloud errors.

dimas-b replied:
I mean that the logic seems to be overly complex when boolean expressions are combined with type-based switches in the same method... but I do not feel too strongly about this... it's a "nit" comment :)

dimas-b replied:
nit: maybe refactor storage exception handling into a separate (private) method?

dimas-b commented on service/common/src/main/java/org/apache/polaris/service/exception/IcebergExceptionMapper.java:
```diff
@@ -144,4 +112,56 @@ public static boolean containsAnyAccessDeniedHint(String message) {
+          };
+
+      if (300 <= httpCode && httpCode <= 499) {
+        return httpCode;
```
I'm not sure that (for example) a 301 (Moved Permanently) response at the storage level is correct to propagate to the Polaris client "as is".

andrew4699 replied:
I thought a bit about this and originally was turning these all into 400s, but landed on this for a few reasons:
1. Polaris is effectively acting as a proxy to their storage service, so it's not unreasonable that it would return an error close to what the storage service returns.
2. Storage credentials/paths are usually configured at a different point-in-time than the request that fails, so calling the subsequent request a "bad request" isn't so accurate.
3. Can make it easier for users to distinguish errors 
4. Still maintains the spirit of errors in the 3xx/4xx range

That said, if we want to follow the REST Catalog [spec](https://github.com/apache/iceberg/blob/main/open-api/rest-catalog-open-api.yaml) more closely, I propose we can do the following:
- 3xx => 400
- 401 => 401
- 403 => 403
- 404 => 404
- Other 4xx => 400

dimas-b replied:
1. I do not think we should consider Polaris as a proxy for storage. Even though today it loads metadata from S3 (for example) on every request, it does not have to be this way in principle... I believe I saw discussions about storing metadata in the Polaris database.

dimas-b replied:
I believe any 3xx error coming from storage, if it's unhandled by Polaris Core, when it comes to this exception mapper is a 500.

dimas-b replied:
401 on the storage side is still a 500 from the client perspective, IMHO, because it basically means the server is not configured properly.

dimas-b replied:
The 403 -> 403 mapping LGTM, but we probably should not expose the storage-side message to the client for security reasons. The client is authorized to access the Polaris Server, but the client may gain information about secrets between the server and storage that the client is not necessarily authorized to know.

dimas-b replied:
404 at storage level,if it bubbles up to this exception mapper is still a 500, IMHO. The storage requests that Polaris makes should be valid by themselves. If the server accepted a request but could not find the file in storage, that should be handled in core and probably wrapped in another exception (which would them be converted to the appropriate error code here).

andrew4699 replied:
> 401 on the storage side is still a 500 from the client perspective, IMHO, because it basically means the server is not configured properly.

> 404 at storage level,if it bubbles up to this exception mapper is still a 500, IMHO. The storage requests that Polaris makes should be valid by themselves. If the server accepted a request but could not find the file in storage, that should be handled in core and probably wrapped in another exception (which would them be converted to the appropriate error code here).

5xx errors should be reserved for incorrect actions the server is taking with the information it has, rather than correct actions taken with invalid client-provided information. Given that the client is responsible for providing the catalog's storage config and actually maintaining its validity (ie not tampering with permissions/files in S3), I think a 401 or 404 are client issues. At a minimum, they are issues the client can attempt to resolve.

> I believe any 3xx error coming from storage, if it's unhandled by Polaris Core, when it comes to this exception mapper is a 500 from the Polaris client perspective. The server basically failed to interact with storage and there's no client-side data that is a factor here.

Cloud Storage can return 3xx when the region is set incorrectly, the path is invalid, and when files are moved. The region and path are both provided by the client when creating the catalog. I don't think it's so simple for the server to handle redirects.

dimas-b replied:
> Cloud Storage can return 3xx when the region is set incorrectly, the path is invalid, and when files are moved. [...]

Interesting :thinking: Still, the response here goes to the Catalog client. This client is most cases is not involved in configuring the catalog. Why should it get a 3xx or 4xx error? 

For a 4xx error the server ought to prove that the client's request is "bad" is some sense, but here the error is not in the request, but in the configuration owned by the server.

A 3xx response indicates the client should re-sent the request to a different URL. Again, this does not make sense if the reason is catalog (mis-) configuration.

If we want to distinguish these failures from "coding mistakes", I think a 502 (Bad Gateway) response may be appropriate from the HTTP perspective and also provide input into finding the root cause. In this case, I think Polaris should add its own message to the HTTP response to make it clear that it is only propagating the storage error to the client.

dimas-b replied:
> 5xx errors should be reserved for incorrect actions the server is taking with the information it has, rather than correct actions taken with invalid client-provided information.

This matches 500 fairly well, but not all 5xx errors, IMHO.

That said, what is "client-provided information"? In the context of this exception mapper, if the client is the REST Catalog, for example, it did not provide catalog configuration as part of the request it sent.

I suppose it may be necessary to have different exception mappers for the REST Catalog API and the Management API for fine-tuning error responses. WDYT?

dimas-b replied:
(Tangential comment about zone IDs, etc.: Ideally storage config should be validated at catalog creation time. Then, at REST Catalog request time, the server will be able to assume the config is valid).

andrew4699 replied:
> (Tangential comment about zone IDs, etc.: Ideally storage config should be validated at catalog creation time. Then, at REST Catalog request time, the server will be able to assume the config is valid).

Agree this would be a nice feature. It unfortunately wouldn't be sufficient as the information can become invalid after creation, ie due to the user making changes in AWS.

> This matches 500 fairly well, but not all 5xx errors, IMHO.
> 
> That said, what is "client-provided information"? In the context of this exception mapper, if the client is the REST Catalog, for example, it did not provide catalog configuration as part of the request it sent.
> 
> I suppose it may be necessary to have different exception mappers for the REST Catalog API and the Management API for fine-tuning error responses. WDYT?

I think these are currently only relevant to the REST Catalog API as the Management API doesn't currently call cloud storage. You're correct that the client didn't provide the information in this request, but they did in a previous request.

The client/server model is tricky in Polaris. The client/server jointly own a dependency of the server: cloud storage resources. It's the client's responsibility to give the server the correct permissions (to assume the role and access the bucket), and then inform the server of the correct role/bucket name. The server has to expose a way for the client to do that, for example by providing a user/role identifier for the client to give those permissions to.

The difference between 4xx and 5xx is generally whose fault it is. The client owns the bucket, so in absence of a code bug it's generally the client's fault (and responsibility to fix) that the bucket is inaccessible.

> A 3xx response indicates the client should re-sent the request to a different URL. Again, this does not make sense if the reason is catalog (mis-) configuration.

I agree the 3xx case is odd. IMO it would make sense to map those to 400 as the solution probably isn't a simple redirect.

> I do not think we should consider Polaris as a proxy for storage. Even though today it loads metadata from S3 (for example) on every request, it does not have to be this way in principle... I believe I saw discussions about storing metadata in the Polaris database.

If that happened then wouldn't this entirely be a non-issue? There will presumably still be requests to storage in which case we should apply some principle.

dimas-b replied:
I proposes this mapping:
Storage -> Client
* 3xx -> 502 : redirects mean the server is not able to properly "proxy" storage data, but this is unlikely to be caused by mistakes in the current request, hence "bad gateway".
* 404 -> 400 : something not found in storage is likely due to mistakes in the submitted data, but we cannot assume the the requested Polaris entity is simply not found.
* 400 -> 400 : bad storage request is likely due to mistakes in the submitted data.
* 401 -> 403 : incorrect authentication settings for storage mean the client is denied access to the related Polaris entity, but we cannot flag that the client did not provide correct authentication in its request to Polaris.
* 403 -> 403 : denied access to storage means the client is denied access to the related Polaris entity.
* 429 -> 429 (throttling)
* default -> 500

collado-mike replied:
I agree with all of these except the `502`. I'd really like to make any user-configuration problems not a 5xx error code. `5xx` generally cause engineers to have to look into the errors and these really ought to be user-configuration errors. I think they ought to be `400` errors.

dimas-b replied:
I support the intention to avoid 5xx errors where those can be attributed to user mistakes. However, I believe the HTTP spec defines responses not from the user's perspective, but from the client's perspective. As such a client making a valid request that did not work because of previous calls (e.g. config changes) possibly made by other clients, is not a "bad request" (400), IMHO.

WDYT about mapping 3xx to 422 (Unprocessable Content) instead?

collado-mike replied:
I think that's a good compromise. I agree that `400` is not really the right status code.

dimas-b commented on service/common/src/main/java/org/apache/polaris/service/exception/IcebergExceptionMapper.java:
```diff
@@ -144,4 +113,71 @@ public static boolean containsAnyAccessDeniedHint(String message) {
+        return Status.BAD_GATEWAY.getStatusCode();
+      }
+      if (Status.Family.familyOf(httpCode) == Status.Family.SERVER_ERROR) {
+        return Status.INTERNAL_SERVER_ERROR.getStatusCode();
```
Since we treat this as a special case, I'd actually convert storage 5xx to client 502 (Bad Gateway). I think this matches HTTP error code definitions exactly.

dimas-b reviewed (commented):
Latest diff LGTM, but I believe @collado-mike suggested changes to the handling of 3xx storage errors.

andrew4699 commented:
Just made some updates:
- Refactored cloud exception mapping
- Updated 3xx to be 422
- Updated 5xx to be 502
- Made 504 and 408 pass-through
- Cleaned up the test

dimas-b commented on quarkus/service/src/test/java/org/apache/polaris/service/quarkus/IcebergExceptionMapperTest.java:
```diff
@@ -37,35 +38,40 @@
+        Map.of(
+            // Map of HTTP code returned from a cloud provider to the HTTP code Polaris is expected
+            // to return
+            302, 422,
```
@collado-mike FYI

dimas-b reviewed (approved):
LGTM. Thanks for bearing with me @andrew4699 :)

andrew4699 commented:
Thanks for the suggestions!
